README.md
# Check/Radio Group


[![Pub](https://img.shields.io/badge/pub-v2.0.0-green)](https://pub.dev/packages/check_radio_group)


A library to easily create radio button and checkbox groups.

Define font size, selection color, position of radios / check and text alignment.


We leave you free to customize the way you like best.

Let's program!

## Getting Started #

In the dependencies: section of your pubspec.yaml, add the following line:

```dart
  check_radio_group: 2.0.0
```

## Usage
```dart
import 'package:check_radio_group/check/check_group.dart';
import 'package:check_radio_group/model/group_style.dart';
import 'package:check_radio_group/model/item_group.dart';
import 'package:check_radio_group/radio/radio_group.dart';
import 'package:flutter/material.dart';

void main() => runApp(MaterialApp(
      title: 'Radio/Check Group',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: SamplePage(),
    ));

class SamplePage extends StatefulWidget {
  @override
  _SamplePageState createState() => _SamplePageState();
}

class _SamplePageState extends State<SamplePage> {
  final List<GroupItem> radioItems = [
    GroupItem(title: 'Radio One'),
    GroupItem(title: 'Radio Two'),
    GroupItem(title: 'Radio Three'),
  ];
  final List<GroupItem> checkItems = [
    GroupItem(title: 'Check One'),
    GroupItem(title: 'Check Two'),
    GroupItem(title: 'Check Three'),
  ];

  GroupItem _selected;

  @override
  void initState() {
    setState(() {
      _selected = radioItems[0];
    });
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    _titleWidget(String title) => Padding(
          padding: const EdgeInsets.all(8.0),
          child: Text(
            title,
            style: TextStyle(fontWeight: FontWeight.bold, fontSize: 16),
          ),
        );

    return Scaffold(
      appBar: AppBar(
        title: Text('Check/Radio Group'),
        centerTitle: true,
      ),
      body: Container(
        margin: EdgeInsets.all(16),
        child: ListView(
          children: <Widget>[
            Center(
              child: _titleWidget('Radio Group Sample'),
            ),
            _titleWidget('Default Radio Group'),
            RadioGroup(
              items: radioItems,
              onSelected: (item) {
                print(item.title);
              },
            ),
            SizedBox(
              height: 32,
            ),
            _titleWidget('Selected Radio Group'),
            RadioGroup(
              items: radioItems,
              selected: _selected,
              onSelected: (item) {
                print(item.title);
              },
            ),
            SizedBox(
              height: 32,
            ),
            _titleWidget('Custom Radio Group Style'),
            RadioGroup(
              items: radioItems,
              selected: _selected,
              style: GroupStyle(
                  activeColor: Colors.red,
                  checkPosition: ListTileControlAffinity.trailing,
                  titleAlign: TextAlign.end,
                  titleStyle: TextStyle(fontSize: 12)),
              onSelected: (item) {
                print(item.title);
              },
            ),
            SizedBox(
              height: 32,
            ),
            Center(
              child: _titleWidget('Check Group Sample'),
            ),
            _titleWidget('Default Check Group'),
            CheckGroup(
              items: radioItems,
              onSelected: (item) {
                print(item.title);
              },
            ),
            SizedBox(
              height: 32,
            ),
            Center(
              child: _titleWidget('Check Group Sample'),
            ),
            _titleWidget('Selected Check Group'),
            CheckGroup(
              items: checkItems,
              style: GroupStyle(
                  activeColor: Colors.red,
                  checkPosition: ListTileControlAffinity.leading,
                  titleAlign: TextAlign.left,
                  titleStyle: TextStyle(fontSize: 12)),
              onSelected: (item) {
                print(item.title);
              },
            ),
          ],
        ),
      ),
    );
  }
}


```


## Demo

<p align="center">
  <img src="https://github.com/caiubitech/check_radio_group/blob/master/radio_sample.png" alt="Radio button group" width="320" height="520"/>
  <img src="https://github.com/caiubitech/check_radio_group/blob/master/checkbox_sample.png" alt="Checkbox Group" width="320"  height="520"/>
</p>


CHANGELOG.md

## [2.0.1]
- Mandatory implementation of the onSelectedItem function removed. 

- Add new function onSelectedItems // As the component is multiple selection, it makes more sense to return the list of selected items

## [1.0.5]
- First version

lib/check/check_group.dart
// Copyright 2020 The CaiubiTech Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
library check_radio_group;

import 'package:check_radio_group/model/group_style.dart';
import 'package:check_radio_group/model/item_group.dart';
import 'package:flutter/material.dart';

class CheckGroup extends StatefulWidget {
  final List<GroupItem> items;
  final Function(List<GroupItem>) onSelectedItems;
  final GroupStyle style;

  const CheckGroup(
      {Key key,
      @required this.items,
      @required this.onSelectedItems,
      this.style})
      : super(key: key);

  @override
  _CheckGroupState createState() => _CheckGroupState();
}

class _CheckGroupState extends State<CheckGroup> {
  GroupStyle _groupStyle;

  List<GroupItem> _selectedItems = [];

  @override
  void initState() {
    setState(() {
      _groupStyle = widget.style ?? GroupStyle();
    });

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.center,
      mainAxisAlignment: MainAxisAlignment.center,
      children: widget.items.map((item) {
        final pos = widget.items.indexOf(item);

        return CheckboxListTile(
          value: item.selected,
          controlAffinity: _groupStyle.checkPosition,
          activeColor: _groupStyle.activeColor ??
              Theme.of(context).toggleableActiveColor,
          title: Text(
            item.title,
            style: _groupStyle.titleStyle,
            textAlign: _groupStyle.titleAlign,
          ),
          subtitle: item?.subTitle != null
              ? Text(
                  item?.subTitle,
                  style: _groupStyle.titleStyle,
                  textAlign: _groupStyle.titleAlign,
                )
              : null,
          onChanged: (value) {
            setState(() {
              item.selected = value;

              if (!item.selected) {
                _selectedItems
                    .removeWhere((value) => value.hashCode == item.hashCode);
              } else {
                _selectedItems.add(item);
              }

              widget.onSelectedItems(_selectedItems);
            });
          },
          key: Key('$pos'),
        );
      }).toList(),
    );
  }
}


lib/radio/radio_group.dart
// Copyright 2020 The CaiubiTech Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
library check_radio_group;

import 'package:check_radio_group/model/group_style.dart';
import 'package:check_radio_group/model/item_group.dart';
import 'package:flutter/material.dart';

class RadioGroup extends StatefulWidget {
  final List<GroupItem> items;
  final GroupItem selected;
  final Function(GroupItem) onSelected;
  final GroupStyle style;

  const RadioGroup(
      {Key key,
      @required this.items,
      @required this.onSelected,
      this.style,
      this.selected})
      : super(key: key);

  @override
  _RadioGroupState createState() => _RadioGroupState();
}

class _RadioGroupState extends State<RadioGroup> {
  GroupItem _selected;
  GroupStyle _groupStyle;

  @override
  void initState() {
    setState(() {
      _selected = widget?.selected ?? null;
    });

    setState(() {
      _groupStyle = widget.style ?? GroupStyle();
    });

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.center,
      mainAxisAlignment: MainAxisAlignment.center,
      children: widget.items.map((item) {
        final pos = widget.items.indexOf(item);

        return RadioListTile(
            key: Key('$pos${item.title}'),
            title: Text(
              item.title,
              style: _groupStyle.titleStyle,
              textAlign: _groupStyle.titleAlign,
            ),
            subtitle: item?.subTitle != null
                ? Text(
                    item?.subTitle,
                    style: _groupStyle.titleStyle,
                    textAlign: _groupStyle.titleAlign,
                  )
                : null,
            controlAffinity: _groupStyle.checkPosition,
            dense: true,
            activeColor: _groupStyle?.activeColor ??
                Theme.of(context).toggleableActiveColor,
            groupValue: widget.items.indexOf(_selected),
            value: pos,
            onChanged: (var index) => setState(() {
                  _selected = item;
                  widget.onSelected(item);
                }));
      }).toList(),
    );
  }
}


example/lib/example.dart

import 'package:check_radio_group/check/check_group.dart';
import 'package:check_radio_group/model/group_style.dart';
import 'package:check_radio_group/model/item_group.dart';
import 'package:check_radio_group/radio/radio_group.dart';
import 'package:flutter/material.dart';

void main() => runApp(MaterialApp(
      title: 'Radio/Check Group',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: SamplePage(),
    ));

class SamplePage extends StatefulWidget {
  @override
  _SamplePageState createState() => _SamplePageState();
}

class _SamplePageState extends State<SamplePage> {
  final List<GroupItem> radioItems = [
    GroupItem(title: 'Radio One'),
    GroupItem(title: 'Radio Two'),
    GroupItem(title: 'Radio Three'),
  ];
  final List<GroupItem> checkItems = [
    GroupItem(title: 'Check One'),
    GroupItem(title: 'Check Two'),
    GroupItem(title: 'Check Three'),
  ];

  GroupItem _selected;

  @override
  void initState() {
    setState(() {
      _selected = radioItems[0];
    });
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    _titleWidget(String title) => Padding(
          padding: const EdgeInsets.all(8.0),
          child: Text(
            title,
            style: TextStyle(fontWeight: FontWeight.bold, fontSize: 16),
          ),
        );

    return Scaffold(
      appBar: AppBar(
        title: Text('Check/Radio Group'),
        centerTitle: true,
      ),
      body: Container(
        margin: EdgeInsets.all(16),
        child: ListView(
          children: <Widget>[
            Center(
              child: _titleWidget('Radio Group Sample'),
            ),
            _titleWidget('Default Radio Group'),
            RadioGroup(
              items: radioItems,
              onSelected: (item) {
                print(item.title);
              },
            ),
            SizedBox(
              height: 32,
            ),
            _titleWidget('Selected Radio Group'),
            RadioGroup(
              items: radioItems,
              selected: _selected,
              onSelected: (item) {
                print(item.title);
              },
            ),
            SizedBox(
              height: 32,
            ),
            _titleWidget('Custom Radio Group Style'),
            RadioGroup(
              items: radioItems,
              selected: _selected,
              style: GroupStyle(
                  activeColor: Colors.red,
                  checkPosition: ListTileControlAffinity.trailing,
                  titleAlign: TextAlign.end,
                  titleStyle: TextStyle(fontSize: 12)),
              onSelected: (item) {
                print(item.title);
              },
            ),
            SizedBox(
              height: 32,
            ),
            Center(
              child: _titleWidget('Check Group Sample'),
            ),
            _titleWidget('Default Check Group'),
            CheckGroup(
              items: radioItems,
              onSelectedItems: (items) {
                items.forEach((item) => print(item.title));
              }
            ),
            SizedBox(
              height: 32,
            ),
            Center(
              child: _titleWidget('Check Group Sample'),
            ),
            _titleWidget('Selected Check Group'),
            CheckGroup(
              items: checkItems,
              style: GroupStyle(
                  activeColor: Colors.red,
                  checkPosition: ListTileControlAffinity.leading,
                  titleAlign: TextAlign.left,
                  titleStyle: TextStyle(fontSize: 12)),
              onSelectedItems: (items) {
                items.forEach((item) => print(item.title));
              }
            ),
          ],
        ),
      ),
    );
  }
}